### PR TITLE
First shot at sidebar accordions and „scroll spy“

### PR DESCRIPTION
**Pull request by fk**
This ports sidebar navigation accordions and „scroll spy“ behavior from the ReactJS docs, with a few adjustments:

* closely stick to the current `yml` format
  * instead of mixing user-defined `id` and `href` values, stick to the existing `link`
  * opt in to setting a „directory“ (think „baseUrl“ for all item links in `yml`) — makes
  * generate an `id` for every `yaml` navigation item

I had to bump gatsby-remark-prismjs from 1.2.23 to 2.0.4 after I — all of a sudden — ran into the `Uncaught TypeError: Cannot read property 'tokenizePlaceholders' of undefined` error.

### CODE REVIEW

File: www/src/components/sidebar/body.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      openSectionHash: {
+        ...openSectionHash,
+        [section.title]: !openSectionHash[section.title],
+      },
+    })
+  }
+
+  _expandAll = () => {
+    const { openSectionHash } = this.state
+    Object.keys(openSectionHash).forEach(k => (openSectionHash[k] = true))
```
Comment (calcsam) [inline www/src/components/sidebar/body.js]: this mutates state directly -- we need to make a copy of the object and then call `setState`

File: www/src/components/sidebar/scroll-sync-body.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  })
+
+  return [].concat(...list)
+}
+
+const _getElementTopOffsetsById = ids =>
+  ids
+    .map(hash => {
+      const element = document.getElementById(hash)
+      if (!element) {
```
Comment (calcsam) [inline www/src/components/sidebar/scroll-sync-body.js]: `return element ? { hash, offsetTop: element.offsetTop - 63, } : null`

File: www/src/components/sidebar/body.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        state.openSectionHash[child.title] =
+          props.defaultActiveSection.title === child.title ? true : false
+      })
+    }
+
+    this.state = state
+  }
+
+  _toggleSection(section) {
+    const openSectionHash = this.state.openSectionHash
```
Comment (calcsam) [inline www/src/components/sidebar/body.js]: `const { openSectionHash } = this.state`

File: www/src/components/sidebar/body.js (hunk 1 above)
Comment (calcsam) [reply www/src/components/sidebar/body.js]: eg `const newOpenSectionHash = { ...openSectionHash }`

File: www/src/components/sidebar/scroll-sync-body.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  handleResize() {
+    this.calculateItemTopOffsets()
+    this.handleScroll()
+  }
+
+  handleScroll() {
+    const { itemTopOffsets } = this.state
+    const item = itemTopOffsets.find((itemTopOffset, i) => {
+      const nextItemTopOffset = itemTopOffsets[i + 1]
+      if (nextItemTopOffset) {
```
Comment (calcsam) [inline www/src/components/sidebar/scroll-sync-body.js]: a `return nextItemTopOffset ? .. : ...` ternary more clear here

**Comment by calcsam** [comment]
Wow, this looks great! Super exciting to see us laying the foundation of the new IA! The scrollSync is really awesome, the new highlighting tab is really neat

Some questions:
* When you open an accordion, and then click another page in a different accordion, your opened accordion closes. Is it possible to do anything about this? May not be since it's a route transition, we may have to wipe accordion state or do something super hacky (`<Link>` with `preventDefault()` and then a `navigateTo`?)
* Format of second-level items. How do we want to present these? Standard `ul` display with `-` as the list item comes to mind for me, but mostly wanted to ask the question. 
image
* It's easy see how we could expand the Docs section to add in new Docs sections. How will this work on the tutorial section? Main question is -- how will subsequent tutorials be discoverable, since everything we put after the main tutorial series will be so far below the fold? Collapse everything by default? Nest everything under a `Introduction to Gatsby` section that is opened by default on the `/tutorial` route, that can be fully collapsed with one click? Etc.

**Comment by fk** [comment]
Thank you so much for reviewing @calcsam, this is incredibly helpful!

> When you open an accordion, and then click another page in a different accordion, your opened accordion closes. Is it possible to do anything about this? May not be since it's a route transition, we may have to wipe accordion state or do something super hacky (<Link> with preventDefault() and then a navigateTo?)

Yes, this is my major UX worry along with the active item not being scrolled into view if below the fold, too. I think I can handle the latter.
Regarding your question, I thought about dropping the "auto-close non-active accordion" behavior altogether in favor of saving the expanded/collapsed state of the accordions in local storage or a cookie. If a user had all accordions expanded, he would be able to return to that state even when starting a new "browsing session". We'd only override the saved state regarding the default section: When navigating to /docs/, the "Quick Start" section should always be expanded so the initially selected navigation item is always visible, even if the user previously chose to collapse that accordion.

> Format of second-level items. How do we want to present these? Standard ul display with - as the list item comes to mind for me, but mostly wanted to ask the question.

I felt just indention would be enough, but I can see how adding a "bullet"-like indicator might be helpful, esp. with multi-line items. FWIW I think the children of "How Gatsby works with…" (from [here](https://github.com/gatsbyjs/rfcs/pull/5#issuecomment-395915639)) worked quite well with just the indention, but that might be due to their parent item ending with an ellipsis:

sidebar clickpath auto-expanding first-level groups

**Comment by fk** [comment]
Superseded by https://github.com/gatsbyjs/gatsby/pull/6153.

**Pull request closed without merging**